nRF53: HCI driver & sample

kapi-no opened the pull request:
- Bluetooth HCI over OpenAMP sample
- Bluetooth HCI driver over OpenAMP

kl-cruz committed:
drivers: ipm: add nRFx IPM driver

This commit contributes the nRFx IPM driver.

Signed-off-by: Karol Lasończyk <karol.lasonczyk@nordicsemi.no>

arnopo committed:
scripts: dts: Add new chosen declaration for the IPC shared memory

The interprocessor communication can be based on shared memory.
Allow to declare this memory as with a generic name derived from
chosen declaration.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@st.com>

Commit:
soc: arm: nordic: adding ipc aliases for nRF53

This commit adds ipc-0 aliases in the DTS framework
for nRF5340 Application and Network MCU.

Signed-off-by: Emil Obalski <emil.obalski@nordicsemi.no>

kapi-no commented on samples/bluetooth/hci_open_amp/src/main.c:
```diff
@@ -0,0 +1,469 @@
+#define LOG_MODULE_NAME hci_open_amp
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#define SHM_START_ADDR      (0x20010000 + 0x400)
```
Is there a DTS symbol for shared memory?

carlescufi replied:
@galak ^^

ioannisg replied:
Yes, @kapi-no there is DTS information in the nrf5340 DTS structure that points to the shared memory.

kapi-no replied:
Yes, I see - there is ```sram0_shared``` DT node, but I don't see any alias / define generated for it

kapi-no commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -30,6 +30,17 @@ config BT_H5
+	select IPM
+	select IPM_NRFX
+	select NRFX_IPC
+	select IPM_MSG_CH_1_ENABLE
```
I would like to configure channel direction here. However, this syntax does not work

```
select IPM_MSG_CH_0_TX
select IPM_MSG_CH_1_RX
```

since IPM_MSG_CH_0_TX is a ```choice``` item. How can I work around this?

SebastianBoe replied:
Could you do what we did with choosing a libc implementation?

Add a layer of indirection by having the user choose something that affected the default choice.

https://github.com/zephyrproject-rtos/zephyr/blob/master/lib/libc/Kconfig#L17

kapi-no replied:
Unfortunately, the choice option is defined within a Kconfig template file (the template file depends on channel number), so it would be messy to have it there (if possible at all)

https://github.com/zephyrproject-rtos/zephyr/blob/9523043d97ce6ed895a02185378971c67e342846/drivers/ipm/Kconfig.nrfx_ipc_channel

kapi-no replied:
It is also impossible to give a parameterized name to the choice option:
```choice IPM_MSG_CH_$(nrfx_ipc_num)_DIR```
Kconfig python parser scripts cannot handle that.

With name, I could define the default in the Kconfig board file:
```
choice IPM_MSG_CH_0_DIR
	default IPM_MSG_CH_0_TX if BT_NRF_OPEN_AMP
endchoice
```

kapi-no reviewed (commented):
Highlighting a few unresolved issues

carlescufi commented:
CC @alexandru-porosanu-nxp

carlescufi reviewed (changes requested):
A quick naming change request.

Both the driver and the sample should be named `hci_rpmsg`. This is because RPMsg is the transport, and you don't need OpenAMP on the other side, you can use any other RPMsg implementation instead.

jhedberg commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,465 @@
+				bt_recv_prio(buf);
+			} else {
+				bt_recv(buf);
+			}
```
The whole idea of the `bt_recv_prio()` vs. `bt_recv()` separation is that they do *not* happen from the same context (thread). This is particularly important when you're implementing the RX thread on the HCI driver side rather than letting the host do it. In this case `bt_recv_prio()` should probably be used by whatever context is feeding `rx_queue` and here in the rx_thread you'd only use `bt_recv()`

jhedberg replied:
Actually, you're not selecting BT_RECV_IS_RX_THREAD in your Kconfig definition but still creating your own RX thread inside the driver? You need to decide whether you want to keep the thread inside the driver or the host (or perhaps both, which would have a substantial memory overhead). As long as you don't select BT_RECV_IS_RX_THREAD it is actually safe to call the recv() APIs like you're doing since the host-side will take care of splitting processing into independent threads, however if you did mean to select BT_RECV_IS_RX_THREAD then you need to separate them inside the driver.

joerchan replied:
> Actually, you're not selecting BT_RECV_IS_RX_THREAD in your Kconfig definition but still creating your own RX thread inside the driver? 

That `BT_RECV_IS_RX_THREAD` option is still poorly named, but the documentation is better.
```

carlescufi replied:
Isn't that the other way around? The host has its own RX thread when `BT_RECV_IS_RX_THREAD` is *not* enabled

jhedberg replied:
> Isn't that the other way around? The host has its own RX thread when BT_RECV_IS_RX_THREAD is not enabled

Yes. The intention of the current option name is to say "The bt_recv() call happens in the Bluetooth RX thread", which then implies that there's no need for the host to have such a thread. I'm open to better names if you think it's confusing.

jhedberg replied:
> Since we call it from the same priority context, we need to select the BT_RECV_IS_RX_THREAD option.

But do you need to call it from the same context? What's the reason there's something called the RX thread inside the driver, and that thread uses the Bluetooth RX thread Kconfig options like BT_RX_THREAD_SIZE. That makes it very much look like "the" Bluetooth RX thread :smile:

So, do you need a separate thread like this in the driver? If not, just call `bt_recv()` and `bt_recv_prio()` directly from the context that currently feeds the `rx_queue` and let the host take care of the details. If you do need the separate thread, move the `bt_recv_prio()` like I suggested and select the `BT_RECV_IS_RX_THREAD` option.

jhedberg replied:
> Since we call it from the same priority context, we need to select the BT_RECV_IS_RX_THREAD option.

Like Carles said it's actually the opposite. My previous comment was based on imagining a "don't" in the above quote 😄

joerchan replied:
> Isn't that the other way around? The host has its own RX thread when `BT_RECV_IS_RX_THREAD` is **not** enabled

Yes, sorry. Messed it up.

In h4.c bt_recv is called from thread. and bt_recv_prio is called from interrupt.
In the hci_driver.c bt_recv is called from a thread, and bt_recv_prio is called from a different thread with higher priority.
In open_amp_nrf.c bt_recv and bt_recv_prio is called from the same thread.

So h4 and hci_driver both fulfull the requirement of sending from different priority contexts, and can select BT_RECV_IS_RX_THREAD.
This one cannot select it because it it calls them from the same priority context.

So then this is correct.

jhedberg replied:
> So then this is correct.

It's not wrong from a HCI driver API contract perspective, but my question still remains: is this "RX thread" in the driver actually needed, and why should it have its stack-size determined by the Kconfig option when it never reaches the application level (that's why the normal RX thread size is user-configurable).

joerchan replied:
Thank you for clarifying. I completely misunderstood your comment. And I think you have a good point (assuming I have understood it correctly :P).

@kapi-no Have you considered getting rid of the TX thread here and calling bt_recv and bt_recv_prio directly from `bt_open_amp_rx`. In `bt_recv` we are basically doing the same thing as what is being done here (see here: https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L4866). That would mean that `bt_recv_prio` would be processed in the context of `bt_open_amp_rx`, which seem to be interrupt context.
So I guess the real question is if there is any downside to calling `bt_recv_prio` directly from interrupt context (there are other HCI drivers that do exactly that).

jhedberg commented on drivers/ipm/Kconfig:
```diff
@@ -6,6 +6,21 @@ menuconfig IPM
+	  Enable this for the first CPU that initializes IPM.
+	  Sets up the initial interrupt mask and clears out all
+	  channels. Should be turned on for one CPU only.
+
```
What's this about? I don't think we've had Quark SE support in the tree for quite awhile.

carlescufi replied:
Yep, just get rid of all Quark-related changes @kapi-no

ioannisg replied:
Something went really wrong in the rebase

kapi-no replied:
Yes, this is leftover from rebasing

SebastianBoe commented on drivers/ipm/Kconfig.nrfx:
```diff
@@ -0,0 +1,39 @@
+if IPM_NRFX
+
+nrfx_ipc_num = 0
+source "drivers/ipm/Kconfig.template.nrf_ipc_channel"
```
rsource?

masz-nordic replied:
Fixed

SebastianBoe commented on drivers/ipm/ipm_nrfx_ipc.c:
```diff
@@ -0,0 +1,253 @@
+		    &vipm_nrf_##_idx##_driver_api);
+
+#define VIPM_DEVICE(_idx, _)						\
+	COND_CODE_1(IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_ENABLE),	\
```
Nice, I haven't seen anyone else figure out how to do this until now.

SebastianBoe commented on drivers/ipm/ipm_nrfx_ipc.c:
```diff
@@ -0,0 +1,253 @@
+};												\
+												\
+DEVICE_AND_API_INIT(vipm_nrf_##_idx , "IPM_"#_idx,						\
+		    vipm_nrf_##_idx##_init, NULL, NULL,						\
```
vipm_nrf_##_idx##_init

We should do as little code generation as possible. Co-incidentally,

vipm_nrf_##_idx##_init is the same for all _idx. So we can have vipm_nrf_n_init, and re-use it for _idx.

masz-nordic replied:
Fixed

SebastianBoe commented on drivers/ipm/ipm_nrfx_ipc.c:
```diff
@@ -0,0 +1,253 @@
+{												\
+	if (!IS_ENABLED(CONFIG_IPM_MSG_CH_##_idx##_ENABLE)) {					\
+		LOG_ERR("IPM_" #_idx " is not enabled");					\
+		return -EINVAL;									\
```
Isn't this code-path impossible due to https://github.com/zephyrproject-rtos/zephyr/pull/20467/files#diff-ec888dfc90d8c3eb9142711f8e6a3281R226 ?

masz-nordic replied:
Right, removed

SebastianBoe commented on drivers/ipm/Kconfig.template.nrf_ipc_channel:
```diff
@@ -0,0 +1,21 @@
+config IPM_MSG_CH_$(nrfx_ipc_num)_TX
+	bool "TX"
+
+endchoice
```
Would changing RX to a bool, and having

config IPM_MSG_CH_$(nrfx_ipc_num)_TX
  default ! IPM_MSG_CH_$(nrfx_ipc_num)_RX

help with your issues with choice?

kapi-no replied:
Do you mean something along these lines?

```
menu "IPM Message Channel [$(nrfx_ipc_num)] configuration"

config IPM_MSG_CH_$(nrfx_ipc_num)_ENABLE
	bool "Enable IPM Message Channel $(nrfx_ipc_num)"

config IPM_MSG_CH_$(nrfx_ipc_num)_RX
	bool "IPM Message RX Channel"
	depends on IPM_MSG_CH_$(nrfx_ipc_num)_ENABLE

config IPM_MSG_CH_$(nrfx_ipc_num)_TX
	bool "IPM Message TX Channel"
	depends on IPM_MSG_CH_$(nrfx_ipc_num)_ENABLE
	default ! IPM_MSG_CH_$(nrfx_ipc_num)_RX

endmenu
```

Yes, getting rid of choice option solves issues with choice :) and I can simply select proper configuration.

SebastianBoe replied:
I think RX and TX could both get enabled with that configuration. I was thinking of having one option promptless to prevent this. But that might not allow you to select from other options as you want.

So I don't really see a clean solution here.

kapi-no replied:
Yes, I am going to use my suggestion in this PR. It could be improved in other PR if we find a cleaner way of doing it

SebastianBoe commented on soc/arm/nordic_nrf/nrf53/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,13 @@
+
+if SOC_SERIES_NRF53X
+
+source "soc/arm/nordic_nrf/nrf53/Kconfig.defconfig.nrf53*"
```
rsource

ioannisg commented:
To start with,  I am sending a patch to add code owners for the Nordic IPM driver

jhedberg commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,465 @@
+	if (buf) {
+		BT_DBG("Remaining: %d", remaining);
+		net_buf_add_mem(buf, data, remaining);
+		net_buf_put(&rx_queue, buf);
```
Coming back to my comment regarding the driver-internal RX thread (which wasn't answered yet, btw), my point was that why can't you just call `bt_recv()`/`bt_recv_prio()` here and forget about the rx_queue and a separate thread? `bt_recv()` will take care of deferring to the host-side RX thread since that gets enabled due to the way this driver is defined in Kconfig.

joerchan replied:
I agree. (commented longer answer in the original comment that wasn't answered.).

kapi-no replied:
Sorry for the late response, but I was offline

@jhedberg, @joerchan 
Yes, it seems that the RX thread is redundant in this case. I was too heavily inspired by h4 implementation of HCI driver, which contained a lot more processing logic (HCI packet assembly, etc.).

I also missed the fact that the context switching (to another (Host) thread) happens right after bt_recv() call like @joerchan pointed out:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/hci_core.c#L4866

I`m going to send a patch fixing it in a second.

kapi-no replied:
done, please take a look

ioannisg committed:
CODEOWNERS: add code owners for nRFx IPM driver

We are adding code owners for the Nordic nRFx IPM
driver files.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

joerchan commented on samples/bluetooth/hci_open_amp/src/main.c:
```diff
@@ -0,0 +1,456 @@
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return 1 << VIRTIO_RPMSG_F_NS;
```
1U in case you wish to bit-shift into last position.

jhedberg replied:
Or rather `BIT(VIRTIO_RPMSG_F_NS)`

kapi-no replied:
fixed

jhedberg reviewed (changes requested):
One thing I just noticed is missing: you need to add a new `default` entry to the `BT_HCI_RESERVE` definition in `subsys/bluetooth/host/Kconfig`. This ensures that you can safely prepend one byte to the buffers for the packet type. Not doing this may lead to a buffer underrun when you do `net_buf_push_u8()`.

kapi-no commented:
> One thing I just noticed is missing: you need to add a new `default` entry to the `BT_HCI_RESERVE` definition in `subsys/bluetooth/host/Kconfig`. This ensures that you can safely prepend one byte to the buffers for the packet type. Not doing this may lead to a buffer underrun when you do `net_buf_push_u8()`.

Thanks for pointing it out, I use `net_buf_push_u8()` function only in the HCI sample so I added an additional entry in the prj.conf.

kapi-no commented:
> 
> 
> A quick naming change request.
> 
> Both the driver and the sample should be named `hci_rpmsg`. This is because RPMsg is the transport, and you don't need OpenAMP on the other side, you can use any other RPMsg implementation instead.

I changed names according to your suggestion, please take a look

joerchan commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,454 @@
+		return NULL;
+	}
+
+	buf = net_buf_alloc(&acl_tx_pool, K_NO_WAIT);
```
Dropping ACL packets seems like a bad idea. We cannot use K_FOREVER if this is in interrupt context. So we need some way to handle this problem.
An alternative is to implement HCI flow control, so that the controller does not send the host more data or events that there are buffers.
This comment goes for all instances of K_NO_WAIT in this PR as well.

kapi-no replied:
Well, I`m relying on the logic that is in the ```bluetooth/controller/hci/hci_driver``` and CONFIG_BT_HCI_ACL_FLOW_CONTROL is set. My buffer pool matches CONFIG_BT_CTLR_TX_BUFFERS. I think this should be enough

carlescufi replied:
This maps directly to `CONFIG_BT_CTLR_TX_BUFFERS` so I don't think there's an issue there. It's one-to-one with it and the buffer is consumed immediately upon call to `bt_send()`. The Host is anyway not allowed to send more until a num complete is issued, so I see no problem here.

jhedberg commented:
> Thanks for pointing it out, I use net_buf_push_u8() function only in the HCI sample so I added an additional entry in the prj.conf.

@kapi-no also see #20622 and my comment there. The reason the hci_raw bug hasn't been noticed is that most commonly the `samples/bluetooth/hci_*` samples are used together with the native controller, which sets the reserved headroom to 0. Any thoughts on removing this Kconfig option and making it an integer member of the bt_hci_driver struct instead? (I'm thinking that might help avoid forgetting to update the Kconfig definition on the host side)

jhedberg commented:
>  The reason the hci_raw bug hasn't been noticed is that most commonly the `samples/bluetooth/hci_*` samples are used together with the native controller, which sets the reserved headroom to 0.

Actually, this *would* be an issue for this PR, since the declared reserved headroom is for the HCI driver, *not* the other transport that the hci_raw application implements. So for the case of the RPMsg-based transport when the underlying driver (native controller) uses 0 headroom, you would need to make sure that your sample app does the right thing when declaring the buffer pool and allocating buffers.

kapi-no commented:
>Actually, this would be an issue for this PR, since the declared reserved headroom is for the HCI driver, not the other transport that the hci_raw application implements. So for the case of the RPMsg-based transport when the underlying driver (native controller) uses 0 headroom, you would need to make sure that your sample app does the right thing when declaring the buffer pool and allocating buffers.

Do you mean that I should reserve 1 byte headroom with net_buf_reserve before pushing to it? Or is there something more to it?

jhedberg commented:
>>  The reason the hci_raw bug hasn't been noticed is that most commonly the `samples/bluetooth/hci_*` samples are used together with the native controller, which sets the reserved headroom to 0.

> Actually, this *would* be an issue for this PR, since the declared reserved headroom is for the HCI driver, *not* the other transport that the hci_raw application implements. So for the case of the RPMsg-based transport when the underlying driver (native controller) uses 0 headroom, you would need to make sure that your sample app does the right thing when declaring the buffer pool and allocating buffers.

The only solution for this (i.e. the headroom requirement of the hci_raw transport, implemented by an app) I can think of right now is to introduce a new Kconfig variable that `depends on BT_HCI_RAW`, which the application (like hci_rpmsg) would set in its prj.conf to whatever it needs for the headroom. `hci_raw.c` would then use `MAX(CONFIG_BT_HCI_RESERVE, CONFIG_BT_HCI_RAW_RESERVE)` as the headroom for buffers.

jhedberg commented:
> Do you mean that I should reserve 1 byte headroom with net_buf_reserve before pushing to it? Or is there something more to it?

@kapi-no the buffers you get will have a max payload of CONFIG_BT_RX_BUF_SIZE, *excluding* any HCI transport headers. So if the HCI transport needs something for itself then the stack (`hci_core.c` and `hci_raw.c`) needs to know about this extra headroom requirement. You could call net_buf_reserve() but if the buffer definitions didn't take the extra headroom into account you'd be eating into the part that's intended for non-HCI transport sage. See my previous comment about introducing a new Kconfig that a hci_raw-using app would set if it needs extra headroom.

kapi-no commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,423 @@
+	}
+
+	io = metal_device_io_region(device, 0);
+	if (!io) {
```
fixed

kapi-no commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,423 @@
+	}
+
+	if (*remaining != hdr.len) {
+		BT_ERR("ACL payload length is not correct");
```
fixed

kapi-no commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,423 @@
+	*remaining -= sizeof(hdr);
+
+	if (*remaining != hdr.len) {
+		BT_ERR("Event payload length is not correct");
```
fixed

kapi-no commented on drivers/bluetooth/hci/open_amp_nrf.c:
```diff
@@ -0,0 +1,423 @@
+	}
+
+	if (buf) {
+		BT_DBG("Remaining: %d", remaining);
```
fixed

Commit:
Bluetooth: hci_raw: Add support for specifying buffer headroom

The HCI transport implemented by an application using the HCI raw
interface may have its own buffer headroom requirements. Currently the
available headroom gets completely determined by the selected HCI
driver. E.g. most of the time this is the native controller driver
which doesn't reserve any headroom at all.

To cover for the needs of HCI raw users, add a new Kconfig variable
for the apps to set to whatever they need. Correspondingly, use the
maximum of the HCI driver and HCI raw headroom requirements for the
buffer pool definitions and the headroom initializations.

Signed-off-by: Johan Hedberg <johan.hedberg@intel.com>

kapi-no commented:
@jhedberg, I included your commit from #20630 and used the ```BT_HCI_RAW_RESERVE``` configuration in ```hci_rpmsg``` sample.

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,426 @@
+static const struct bt_hci_driver drv = {
+	.name		= "RPMsg",
+	.open		= bt_rpmsg_open,
+	.send		= bt_rpmsg_send,
```
I think you're missing here `.bus = BT_HCI_DRIVER_BUS_IPM,`

kapi-no replied:
yes, thanks

jhedberg commented on drivers/ipm/ipm_nrfx_ipc.h:
```diff
@@ -0,0 +1,88 @@
+		[12] = (1UL << 12),
+		[13] = (1UL << 13),
+		[14] = (1UL << 14),
+		[15] = (1UL << 15),
```
These could all be using the `BIT()` macro instead of the custom `1ULL << n` construct.

masz-nordic replied:
Done

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,426 @@
+	struct net_buf *buf;
+
+	if (remaining < sizeof(hdr)) {
+		BT_ERR("Not enought data for event header");
```
minor typo: s/enought/enough/

kapi-no replied:
fixed

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,426 @@
+		BT_ERR("No available ACL buffers!");
+	}
+
+	if (remaining != hdr.len) {
```
`remaining != sys_le16_to_cpu(hdr.len)`. That said, you'd then do the conversion twice, so if you want to avoid that some redesign is needed. Note that for ACL the length header is 16 bits, but for HCI events it's 8 bits (so no conversion necessary there)

kapi-no replied:
Actually, the line of code with the second conversion is not needed

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,426 @@
+		return err;
+	}
+
+	return 0;
```
This could all be replaced with a single line `return bt_hci_driver_register(&drv)`, unless you plan to put more stuff into this function in the future.

kapi-no replied:
simplified

jhedberg commented on drivers/ipm/ipm_nrfx_ipc.c:
```diff
@@ -0,0 +1,252 @@
+	while (event_mask) {
+		u8_t event_idx = __CLZ(__RBIT(event_mask));
+
+		event_mask &= ~(1UL << event_idx);
```
`BIT(event_idx)`

kl-cruz replied:
It could be changed in the 34 line as well.

masz-nordic replied:
Done

jhedberg commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+		buf = net_buf_get(&rx_queue, K_FOREVER);
+		err = hci_rpmsg_send(buf);
+		if (err) {
+			LOG_ERR("Failed to send");
```
To help debugging, perhaps make this `"Failed to send (err %d)", err`

kapi-no replied:
done

jhedberg reviewed (changes requested):
Looks generally good, but I found some minor issues that you may want to fix

carlescufi reviewed (approved):
LGTM. Note that there's a couple of possible improvements but the overall approach is correct, and mimics the logic in the other HCI drivers. I think there's a certain amount of optimizations we could get in from code carried over from the `hci_uart` sample, but those are very minimal and it is not completely clear that they would actually contribute to better footprint.

carlescufi commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,426 @@
+#define VRING_ALIGNMENT     4
+#define VRING_SIZE          16
+
+#define VDEV_STATUS_ADDR    0x20010000
```
@galak as part of the TSC call we discussed making the hardware configuration a bit more flexible, perhaps splitting it out into a separate file. Are we talking exclusively about these numbers?

carlescufi replied:
@kapi-no could you move lines 31-247 approximately into a separate file, named for example `rpmsg_nrf53.c` in the same folder, so that we move the nRF53-specific side of things to its own file?

kapi-no replied:
@galak, @carlescufi, I split openAMP related code and configuration to separate nrf53 file

galak reviewed (changes requested):
split rpmsg.c into nrf53 specific bits from rpmsg bits.  So something like rpmsg_nrf53.c for the rpmsg init, shared memory addr, etc.

dbkinder commented on samples/bluetooth/hci_rpmsg/README.rst:
```diff
@@ -0,0 +1,40 @@
+This sample can be found under :zephyr_file:`samples/bluetooth/hci_rpmsg`
+in the Zephyr tree.
+
+To use this application, you need a board with a Bluetooth controller
```
These sample instructions feel incomplete. Can you provide a specific board you've tested with, and show the build steps and output when running this sample?  Exactly what does the sample do?  How do you set up a separate Bluetooth HCI RPMsg peer?  You list one Kconfig option that might need adjusting, what are the others?

grochu replied:
This sample is supposed to work on `nrf5340_dk_nrf5340_cpunet` like _sample.yaml_ suggests in "platform_whitelist". This is the board it was tested with.
In general, this sample is really a standalone BLE Controller compatible with HCI driver and uses RPMsg transport with the IPM driver as a backend. As such, it is only functional with a counterpart BLE sample which has the BLE Host stack and uses the Bluetooth HCI driver with RPMsg transport (also part of this PR).
Does it answer your question?

ioannisg commented on drivers/ipm/Kconfig:
```diff
@@ -67,6 +67,20 @@ config IPM_MHU
+
+config IPM_NRF_SINGLE_INSTANCE
+	bool "Single instance of IPM device"
+	default n
```
@masz-nordic add some help text here

masz-nordic replied:
Done

ioannisg replied:
thanks

arnopo commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+
+/* Configuration defines */
+
+#define SHM_START_ADDR      (0x20010000 + 0x400)
```
Could you have  alook to this PR https://github.com/zephyrproject-rtos/zephyr/pull/17553 ?
An alternative is proposed to use  "chosen" declaration, in order to avoid this dependence on the platform in code.

kapi-no replied:
thanks for the tip, I cherry-picked commit to have ipm defines generated from dts

ioannisg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -30,6 +30,19 @@ config BT_H5
+	bool "RPMsg HCI"
+	select IPM
+	select IPM_NRFX
+	select NRFX_IPC
```
@kapi-no this selection is redundant here, isn't it? I think IPM_NRFX selects NRFX_IPC

ioannisg replied:
As is, BT_RPMSG seems Nordic-specific, but this is not reflected in the nomenclature; has this been attended somehow?

kapi-no replied:
removed redundant NRFX_IPC

kapi-no replied:
The select part of the config is placed now under BT_RPMSG_NRF53 Kconfig option

arnopo commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	LOG_INF("Got callback of id %u", id);
+	virtqueue_notification(vq[1]);
```
There are some lock in OpenAMP to protect resource. Calling this function in IRQ context can lead to deadlock.

kapi-no replied:
Now it executes in workqueue context.

arnopo commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+
+	rdev = rpmsg_virtio_get_rpmsg_device(&rvdev);
+
+	err = rpmsg_create_ept(&ep, rdev, "k", RPMSG_ADDR_ANY,
```
could you give an explicit name for the service ( e.g. "hci_bt")? it is an example, but it is a good practice to give a name in relation to the feature implemented, can help for debugging.

kapi-no replied:
yes, thank you, I changed it "bt_hci"

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,192 @@
+
+	if (remaining != sys_le16_to_cpu(hdr.len)) {
+		BT_ERR("ACL payload length is not correct");
+		return NULL;
```
There's a buffer leak here. You should be doing `net_buf_unref(buf)` before returning

kapi-no replied:
fixed

jhedberg commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,192 @@
+static inline bool bt_buf_get_prio(struct net_buf *buf)
+{
+	return (bool)(*((u8_t *)net_buf_user_data(buf) + 1));
+}
```
I'm not strictly opposed to having these, but now that you don't have a separate thread and queue anymore you could have made `bt_rpmsg_evt_recv()` return the priority instead of embedding it into the buffer user data, e.g. with an extra `bool *prio` parameter. Another problem with these is that they infringe on the bt_buf namespace, i.e. reading the calling code you might get an impression that they're public functions declared in `include/bluetooth/buf.h`

kapi-no replied:
done

jhedberg commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		LOG_ERR("No available command buffers!");
```
Don't you need to `return NULL;` in this branch? Further down the function there would otherwise be NULL-pointer dereferences, like the `net_buf_add_mem()` call.

kapi-no replied:
fixed

jhedberg commented on samples/bluetooth/hci_rpmsg/src/main.c:
```diff
@@ -0,0 +1,455 @@
+
+		net_buf_add_mem(buf, &hdr, sizeof(hdr));
+	} else {
+		LOG_ERR("No available ACL buffers!");
```
Same here regarding NULL-pointer dereferences later, i.e. look like a missing return statement.

kapi-no replied:
fixed

jhedberg reviewed (changes requested):
Thanks for doing the changes I requested, look much better now! I still found a few issues that look like bugs though.

kapi-no committed:
boards: arm: nrf5340_dk_nrf5340: bind shared memory with ipc shim in dts

Bound shared memory definition with IPC shim for nRF5340 DK.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

arnopo commented on drivers/bluetooth/hci/rpmsg.c:
```diff
@@ -0,0 +1,192 @@
+/*
```
A file with this name already exists in the OpenAMP library, this could be a problem with some compilers (e.g. IAR)....  
Then i don't know if there are rules regarding the file and function namings...
I would suggest to rename it hci_rpmsg.c or rpmsg_hci.c  and function hci_rpmsg_*
But only a recommendation :)

ioannisg replied:
@arnopo , thanks for this note, actually :)
So, I would have no problem with renaming this to hci_rpmsg.c, this would solve your concern.
But then other HCI drivers here are named directly after the underlying transport, e.g. spic.c.
As long as we only care about GCC I see no problem.
@carlescufi your opinion?

carlescufi replied:
I see no problem at all with keeping the current name. I believe if we come to this problem we will need to solve it in a wider fashion across the  Zephyr codebase.

ioannisg replied:
Fair point. Actually @kapi-no had a suggestion to rename to rpmsg_hci.c. I have no strong opinion on this matter, as I said above.

ulfalizer commented on boards/arm/nrf5340_dk_nrf5340/nrf5340_dk_nrf5340_shared_sram_planning_conf.dts:
```diff
@@ -12,6 +12,13 @@
+    chosen {
+      /*
+       * shared memory reserved for the inter-processor communication
+       */
```
DTS supports `//` comments as well by the way.

kapi-no replied:
good to know, but I think that Zephyr sticks to /* */ commenting style (even for one-liners)

ulfalizer commented on drivers/ipm/Kconfig.nrfx:
```diff
@@ -0,0 +1,41 @@
+nrfx_ipc_num = 15
+rsource "Kconfig.nrfx_ipc_channel"
+
+endif #IPM_NRFX
```
Could put a space before `IPM_NRFX`.

kapi-no replied:
done

ulfalizer commented on drivers/ipm/Kconfig:
```diff
@@ -67,6 +67,24 @@ config IPM_MHU
+
+config IPM_NRF_SINGLE_INSTANCE
+	bool "Single instance of IPM device"
+	default n
```
Can skip `default n`, for consistency. `n` is the default default.

That makes the generated docs a bit nicer too.

ioannisg replied:
Agree - default n needs to be removed, @kapi-no

kapi-no replied:
done

ioannisg commented on boards/arm/nrf5340_dk_nrf5340/Kconfig:
```diff
@@ -20,6 +20,16 @@ config BOARD_ENABLE_DCDC_HV
+endchoice
+
+config HEAP_MEM_POOL_SIZE
+	default 8192 if BT_RPMSG
```
@kapi-no this value is not supported  as far as I can see. This seems to be a documentation error, though.
I am just wondering; do you really need that much of heap? Can you live with 4096 and then we update this when we fix the documentation?

kapi-no replied:
Changed to 4096, the test samples seem to be working as well. The configuration could be tweaked later on if this is not an optimal number.

ioannisg replied:
perfect, thanks

ioannisg commented on drivers/bluetooth/hci/CMakeLists.txt:
```diff
@@ -3,5 +3,7 @@
-zephyr_sources_ifdef(CONFIG_BT_STM32_IPM      ipm_stm32wb.c)
+zephyr_sources_ifdef(CONFIG_BT_RPMSG	rpmsg.c)
+zephyr_sources_ifdef(CONFIG_BT_RPMSG_NRF53	rpmsg_nrf53.c)
+zephyr_sources_ifdef(CONFIG_BT_STM32_IPM	ipm_stm32wb.c)
```
I suppose just indentation change in this line...

kapi-no replied:
yes, this is just indentation alignment

ioannisg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -30,6 +30,12 @@ config BT_H5
 	  Three-Wire UART Transport Layer.
 
+config BT_RPMSG
+	bool "RPMsg HCI"
```
Make it "HCI using RPMsg"? Or "HCI over RPMsg transport"?

kapi-no replied:
fixed

ioannisg commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -92,3 +98,18 @@ config BT_SPI_BLUENRG
+
+if BT_RPMSG
+
+config BT_RPMSG_NRF53
```
I would like to see a minor help text here explaining the channel configuration

kapi-no replied:
added short description

ioannisg commented on drivers/ipm/Kconfig:
```diff
@@ -67,6 +67,24 @@ config IPM_MHU
+
+config IPM_NRF_SINGLE_INSTANCE
+	bool "Single instance of IPM device"
+	default n
```
Remove default n - this is not needed

kapi-no replied:
done

carlescufi commented:
@jhedberg @arnopo @dbkinder can you please take another look?

ioannisg reviewed (approved):
Looks pretty good now, approved.
@kapi-no can you take care of the minor issues i have spotted i the last pass?

kapi-no committed:
samples: bluetooth: adding hci_rpmsg sample

This commit contributes a BLE HCI-over-RPMsg sample.

Co-authored-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>
Co-authored-by: Nikodem Kastelik <nikodem.kastelik@nordicsemi.no>

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
drivers: bluetooth: hci: driver based on RPMsg transport

This commit contributes an RPMsg-based transport for BLE HCI.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
boards: nrf5340_dk_nrf5340: enable VS commands if building with BT

When building with support for BLE stack, enable the Vendor Specific
commands for the nRF5340 APP CPU (Application MCU).

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
boards: nrf5340_dk_nrf5340: enable ECC if building with BT

When building with support for BLE stack, enable the BT_ECC
for the nRF5340 NET CPU (Network MCU).

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

kapi-no committed:
boards: nrf5340_dk_nrf5340: use RPMsg as default HCI driver for BLE

When building with support for BLE stack on the nRF5340 APP CPU
(Application MCU), use RPMsg HCI driver by default.

Signed-off-by: Kamil Piszczek <Kamil.Piszczek@nordicsemi.no>

arnopo commented on drivers/bluetooth/hci/rpmsg_nrf53.c:
```diff
@@ -0,0 +1,241 @@
+	/* Since we are using name service, we need to wait for a response
+	 * from NS setup and than we need to process it
+	 */
+	virtqueue_notification(vq[0]);
```
This seems useless. when called remote processor still not ready to send the NS message, so nothing to process in queue.
Application will be blocked on sync_sem waiting rpmsg from the remote side. 
ipm_callback should call the virtqueue_notification that will process message and call the ns_bind_cb

kapi-no replied:
You are right - this is redundant.

kapi-no replied:
fixed

arnopo commented on drivers/bluetooth/hci/rpmsg_nrf53.c:
```diff
@@ -0,0 +1,241 @@
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	BT_DBG("Got callback of id %u", id);
+	virtqueue_notification(vq[0]);
```
Same issue,  processing in interrupt context

kapi-no replied:
I tried to switch context for ```virtqueue_notification``` from ISR to the workqueue context, but oddly enough this triggers MPU fault.

There is a flash write attempt that crashes the application. It occurs inside ```sys_dlist_remove``` in the ```k_sys_work_q``` context. Do you see any potential scenario in which the MPU fault could be caused by OpenAMP library (e.g. pointer corruption)?

kapi-no replied:
However, using a dedicated thread to process ```virtqueue_notification``` seems to work.

arnopo replied:
if you are sending  a message you can take rdev->lock (rpmsg_virtio_send_offchannel_raw). And ipm_callback is called you will treat received messages , try to take rdev->lock a second time(rpmsg_virtio_rx_callback) => deadlock 
yes using a dedicated thread is a good solution

kapi-no replied:
Added a dedicated thread for ```virtqueue_notification```

arnopo commented on drivers/bluetooth/hci/rpmsg_nrf53.c:
```diff
@@ -0,0 +1,241 @@
+}
+
+int bt_rpmsg_platform_init(void)
+{
```
This implementation is working but is not optimized as use one OpenAMP instance for one RPMSg service.
We should be able to share these vrings with some other rpmsg client drivers (refer to https://github.com/zephyrproject-rtos/zephyr/issues/20419 discussion).
 Can probably be addressed later in a separate thread...

ioannisg replied:
Yes, I think so @arnopo , let's take this on separate thread :)

kapi-no commented:
@arnopo, please rereview
